src/wrapper-utils.js

```diff
@@ -0,0 +1,131 @@
+/* global cloneInto, exportFunction, mozProxies */
+// Tests don't define this variable so fallback to behave like chrome
+const hasMozProxies = typeof mozProxies !== 'undefined' ? mozProxies : false
+const globalObj = typeof window === 'undefined' ? globalThis : window
+const getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor
+const functionToString = Function.prototype.toString
+const objectKeys = Object.keys
+const NativeProxy = Proxy
+
+export function defineProperty (object, propertyName, descriptor) {
+    if (hasMozProxies) {
+        const usedObj = object.wrappedJSObject
+        const UsedObjectInterface = globalObj.wrappedJSObject.Object
+        const definedDescriptor = new UsedObjectInterface();
+        ['configurable', 'enumerable', 'value', 'writable'].forEach((propertyName) => {
+            if (propertyName in descriptor) {
+                definedDescriptor[propertyName] = cloneInto(
+                    descriptor[propertyName],
+                    definedDescriptor,
+                    { cloneFunctions: true })
+            }
+        });
+        ['get', 'set'].forEach((methodName) => {
+            if (methodName in descriptor && typeof descriptor[methodName] !== 'undefined') { // Firefox returns undefined for missing getters/setters
+                exportFunction(descriptor[methodName], definedDescriptor, { defineAs: methodName })
+            }
+        })
+        UsedObjectInterface.defineProperty(usedObj, propertyName, definedDescriptor)
+    } else {
+        Object.defineProperty(object, propertyName, descriptor)
+    }
+}
+
+/**
+ * For each property defined on the object, update it with the target value.
+ */
+export function overrideProperty (name, prop) {
+    // Don't update if existing value is undefined or null
+    if (!(prop.origValue === undefined)) {
+        /**
+         * When re-defining properties, we bind the overwritten functions to null. This prevents
+         * sites from using toString to see if the function has been overwritten
+         * without this bind call, a site could run something like
+         * `Object.getOwnPropertyDescriptor(Screen.prototype, "availTop").get.toString()` and see
+         * the contents of the function. Appending .bind(null) to the function definition will
+         * have the same toString call return the default [native code]
+         */
+        try {
+            defineProperty(prop.object, name, {
+                // eslint-disable-next-line no-extra-bind
+                get: (() => prop.targetValue).bind(null)
+            })
+        } catch (e) {
+        }
+    }
+    return prop.origValue
+}
+
+/**
+ * add a fake toString() method to a wrapper function to resemble the original function
+ * @param {*} newFn
+ * @param {*} origFn
+ */
+function wrapToString (newFn, origFn) {
+    if (typeof newFn !== 'function' || typeof origFn !== 'function') {
+        return
+    }
+    newFn.toString = function () {
```

Comment:
I think `x.toString.toString()` will fail, we've had reports for this in the past.

We should probably raise an issue to track.